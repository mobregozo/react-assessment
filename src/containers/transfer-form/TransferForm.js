import React from "react";
import InputForm from "../../components/input-form/InputForm";
import { connect } from "react-redux";
import { transferMoney } from "../../redux/actions";
import Loader from "../../components/loader/loader";
import SubmitButton from "../../components/submit-button/SubmitButton";
import Header from "../../components/header/Header";
import InputBox from "../../components/input-box/InputBox";
import FormBox from "../../components/form-box/FormBox";

class TransferForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      amount: "",
      email: "",
      name: "",
      nameError: true,
      emailError: true,
      amountError: true,
    };

    this.handleChangeName = this.handleChangeName.bind(this);
    this.handleChangeAmount = this.handleChangeAmount.bind(this);
    this.handleChangeEmail = this.handleChangeEmail.bind(this);

    this.handleSubmit = this.handleSubmit.bind(this);
  }

  errorMessage = "Please enter a valid ";

  handleChangeName(event) {
    if (event.target.value.length > 0 && event.target.value.length < 120){
      this.setState({ nameError :false});
    }
    else{
      this.setState({ nameError :true});
    }
    this.setState({ name: event.target.value });
  }

  handleChangeAmount(event) {
    const value = Number(event.target.value);
    if (value && value > 0){
      this.setState({amountError :false});
    }
    else{

      this.setState({amountError :true});
    }
    this.setState({ amount: value});
  }

  handleChangeEmail(event) {
    var re = /\S+@\S+\.\S+/;
    if (re.test(event.target.value)){
      this.setState({emailError : false});
    }
    else{
      this.setState({emailError :true});
    }
    this.setState({ email: event.target.value });
  }

  checkValidation() {
    const availablefunds = this.props.availablefunds >= this.state.amount;
    return availablefunds;
  }

  handleSubmit(event) {
    event.preventDefault();
    if (this.checkValidation()){
      this.props.transferMoney({
        amount: this.state.amount,
        email: this.state.email,
        name: this.state.name
      });
      this.setState({
        amount: "",
        email: "",
        name: "",
        nameError: true,
        emailError: true,
        amountError: true,
      });
    }
    else{
      alert('You do not have enough fund to perform this action');
    }    
    
  }

  render() {
    const { loading } = this.props;
    return (
      <React.Fragment>
        <Header text="Send Money" />

        {loading ? (
          <Loader />
        ) : (
          <FormBox onSubmit={this.handleSubmit} noValidate>
            <InputBox>
              <InputForm
                label="Name"
                type="text"
                errorMessage={this.state.nameError? this.errorMessage+' name' : null}
                value={this.state.name}
                onChange={this.handleChangeName}
              />
              <InputForm
                label="Email Address"
                errorMessage={this.state.emailError? this.errorMessage+' email' : null } 
                type="email"
                value={this.state.email}
                onChange={this.handleChangeEmail}
              />
              <InputForm
                label="Amount"
                errorMessage={this.state.amountError? this.errorMessage+' amount' : null}
                type="number"
                value={this.state.amount}
                onChange={this.handleChangeAmount}
              />
            </InputBox>
            <SubmitButton disabled={this.state.nameError || this.state.emailError || this.state.amountError}/>
          </FormBox>
        )}
      </React.Fragment>
    );
  }
}

const mapStateToProps = ({ transactions, account }) => {
  return {
    loading: transactions.loading,
    availablefunds: account.available
  };
};

export default connect(
  mapStateToProps,
  { transferMoney }
)(TransferForm);
