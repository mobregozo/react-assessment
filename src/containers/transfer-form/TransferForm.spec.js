import React from "react";
import TransferForm from "./TransferForm";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";

// create any initial state needed
const initialState = {
  account: {
    spent: 0,
    available: 3000
  },
  transactions: {
    list: [],
    loading: false
  }
};
// here it is possible to pass in any middleware if needed into //configureStore
const mockStore = configureStore();
let store;

describe('TransferForm', ()=>{
  beforeEach(() => {
    //creates the store with any initial state or middleware needed
    store = mockStore(initialState);
  
  })

  test("Link changes the class when hovered", () => {
    const component = renderer.create(<TransferForm store={store}/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    console.log({tree})
  })
})
