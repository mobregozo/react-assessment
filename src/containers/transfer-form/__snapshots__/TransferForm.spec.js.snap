// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransferForm Link changes the class when hovered 1`] = `
Array [
  <div
    className="sc-gzVnrw dsYuEe"
    size={undefined}
>
    Send Money
</div>,
  <form
    className="sc-dnqmqq fybLen"
    onSubmit={[Function]}
>
    <div
        className="sc-htoDjs cTasum"
    >
        <div
            className="sc-bdVaJa ehXfsk"
        >
            <label
                className="sc-bwzfXH TTgLG"
            >
                Name
            </label>
            <input
                className="sc-htpNat bAmGbS"
                onChange={[Function]}
                type="text"
                value=""
            />
            <div
                className="sc-bxivhb cIBdLZ"
            >
                Please enter a valid  name
            </div>
        </div>
        <div
            className="sc-bdVaJa ehXfsk"
        >
            <label
                className="sc-bwzfXH TTgLG"
            >
                Email Address
            </label>
            <input
                className="sc-htpNat bAmGbS"
                onChange={[Function]}
                type="email"
                value=""
            />
            <div
                className="sc-bxivhb cIBdLZ"
            >
                Please enter a valid  email
            </div>
        </div>
        <div
            className="sc-bdVaJa ehXfsk"
        >
            <label
                className="sc-bwzfXH TTgLG"
            >
                Amount
            </label>
            <input
                className="sc-htpNat bAmGbS"
                onChange={[Function]}
                type="number"
                value=""
            />
            <div
                className="sc-bxivhb cIBdLZ"
            >
                Please enter a valid  amount
            </div>
        </div>
    </div>
    <input
        className="sc-bZQynM bTjdop"
        disabled={true}
        type="submit"
        value="Send"
    />
</form>,
]
`;
